WordPressKit/Authenticator.swift

```diff
@@ -0,0 +1,213 @@
+import Alamofire
+import CocoaLumberjack
+import Foundation
+import WordPressShared
+
+public typealias RequestAuthenticationValidator = (URLRequest) -> Bool
+
+// MARK: - Authenticator
+
+public struct Authenticator {
+    let adapter: RequestAdapter?
+    let retrier: RequestRetrier?
+
+    public init(authenticator: RequestRetrier & RequestAdapter) {
+        adapter = authenticator
+        retrier = authenticator
+    }
+
+    public init(authenticator: RequestRetrier) {
+        adapter = nil
+        retrier = authenticator
+    }
+
+    public init(authenticator: RequestAdapter) {
+        adapter = authenticator
+        retrier = nil
+    }
+}
+
+// MARK: - Token Auth
+
+public struct TokenAuthenticator: RequestAdapter {
+    fileprivate let token: Secret<String>
+    fileprivate let shouldAuthenticate: RequestAuthenticationValidator
+
+    public init(token: String, shouldAuthenticate: RequestAuthenticationValidator? = nil) {
+        self.token = Secret(token)
+        self.shouldAuthenticate = shouldAuthenticate ?? { _ in true }
+    }
+
+    public func adapt(_ urlRequest: URLRequest) throws -> URLRequest {
+        guard shouldAuthenticate(urlRequest) else {
+            return urlRequest
+        }
+        var adaptedRequest = urlRequest
+        adaptedRequest.addValue("Bearer \(token.secretValue)", forHTTPHeaderField: "Authorization")
+        return adaptedRequest
+    }
+}
+
+public extension Authenticator {
+    static func token(_ token: String, shouldAuthenticate: RequestAuthenticationValidator? = nil) -> Authenticator {
+        let authenticator = TokenAuthenticator(token: token, shouldAuthenticate: shouldAuthenticate)
+        return Authenticator(authenticator: authenticator)
+    }
+
+}
+
+// MARK: - Cookie Nonce Auth
+
+public class CookieNonceAuthenticator: RequestAdapter, RequestRetrier {
+    private let username: String
+    private let password: Secret<String>
+    private let loginURL: URL
+    private let adminURL: URL
+    private var nonce: Secret<String>? = nil
+    // If we can't get this to work once, don't retry for the same site
+    // It is likely that there is something preventing us from extracting a nonce
+    private var canRetry = true
+    private var isAuthenticating = false
+    private var requestsToRetry = [RequestRetryCompletion]()
+
+    public init(username: String, password: String, loginURL: URL, adminURL: URL, nonce: String? = nil) {
+        self.username = username
+        self.password = Secret(password)
+        self.loginURL = loginURL
+        self.adminURL = adminURL
+        self.nonce = nonce.map(Secret.init)
+    }
+
+    // MARK: Request Adapter
+
+    public func adapt(_ urlRequest: URLRequest) throws -> URLRequest {
+        guard let nonce = nonce else {
+            return urlRequest
+        }
+        var adaptedRequest = urlRequest
+        adaptedRequest.addValue(nonce.secretValue, forHTTPHeaderField: "X-WP-Nonce")
```

Comment:
`nonce` is only in the constructor for testing purposes, but in the normal flow, it's actually set by the retrier if `startLoginSequence` is successful.

I understand that the adapter and the retrier might look unnecessarily coupled in this first iteration, but the idea was to make this the base for the upcoming auth mechanisms. Most authentication methods (if not all) will require some way to inject something in the request, and to be able to inspect the retry and potentially retry after re-authenticating.

I think I can simplify this a bit by making it a protocol instead of a struct.

> And in the future if we want to pass a different purpose RequestRetrier or RequestAdapter we wouldn't need to change this class. Thus, better interface segregation.

I also wondered about this, but I believe that if we want to do that we'll need to figure out a way to combine adapters and retriers, because you can only assign one to a session.

Reply:
> Most authentication methods (if not all) will require some way to inject something in the request, and to be able to inspect the retry and potentially retry after re-authenticating.

That makes sense. And I don't have a better suggestion currently about how to combine them. I can think of some other ways but can't claim they are any better.